Feature: Add a "Transform" option for paragraph blocks

blixt-io opened the pull request:
## Description
If you tap on the "..." button on a paragraph block, you'll see a Transform option at the top of the list, which will allow you to transform that block to other block-types. Tapping that "Transform" option will transform the paragraph block into a `core/heading` block - in a future pull-request, I'll present a menu of different options for the transformation!

This is a partial implementation of https://github.com/wordpress-mobile/gutenberg-mobile/issues/2814

## How has this been tested?
I've tested this change in the iOS Simulator:
1. Tap the `...` button on a `core/paragraph` block in the mobile Gutenberg editor.
2. There's now a "Transform block" button at the top of the list - tap it!
3. You'll now see a list of transformations that are available - I've excluded `core/column` and `core/group`, since it sounds like those can be confusing.
4. Tap an option, and the block is transformed! There's an info toaster-notification displayed at the top of the screen, too.

## Screenshots <!-- if applicable -->
![CleanShot 2020-11-21 at 22 23 23](https://user-images.githubusercontent.com/72723520/99896741-4ad9fe80-2c48-11eb-9219-0a1b5e1adb82.gif)

## Types of changes
If the `blockType.name === 'core/paragraph'`, then a "Transform block" option is at the top of the `Picker` component. Tapping that option (for now) transitions to a `core/heading` block.

In future pull-requests, I'll add the second Picker where users can select from different destination-blocks.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation.
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

Commit:
Add a "Transform" option for paragraph blocks

If you tap on the "..." button on a paragraph block, you'll see a Transform option at the top of the list. Doesn't do anything yet!

Commit:
Fix-up some linter issues

Commit:
Fix remaining linter issues

Commit:
Use blockName instead of blockType.name

Commit:
Add hard-coded transform to core/heading

Commit:
A few more linter fixes

Commit:
Updates to package-lock.json

I think this is what the checks are upset about - let's see!

chipsnyder reviewed (commented):
Hey Blixit, 
Thanks for your work here so far. It looks like we're making some good progress. 

If you'll permit me, I wanted to start with the PR section, only because this tends to be where I first look when starting a PR review. 

```
## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
```

I see you filled this out with 

> I've tested this change in the iOS Simulator.

To me, this explanation answers the question of where you tested but not necessarily how. We each have a different style (you can see some of the team's [PRs](https://github.com/WordPress/gutenberg/issues?q=label%3A%22Mobile+App+Android%2FiOS%22+is%3Aclosed) if you need inspiration) for how we like to format this, but the things we all try to include are:
The steps necessary to complete a flow, First do A, then B, etc
A test case for each scenario. For you might think of a test case for a paragraph with no text, one with text, etc

Commit:
Implement second-Picker for transforming core/paragraph block (#1)

If you tap on the `...` button for a `core/paragraph` block, you can now transform it into a handful of options, as outlined in https://github.com/wordpress-mobile/gutenberg-mobile/issues/2814

Commit:
Enable transformation between all compatible block types (#2)

The app now uses the function `getPossibleBlockTransformations` to perform any transform on-mobile that we'd perform on web!

Commit:
Extract getMobileBlockTransformations function

Commit:
Add a few unit tests for getMobileBlockTransformations

Commit:
Merge branch 'paragraphTransform-01-add-menu-option' of https://github.com/blixt-io/gutenberg into paragraphTransform-01-add-menu-option

# Conflicts:
#	packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js

Commit:
Merge branch 'master' into paragraphTransform-01-add-menu-option

blixt-io commented on packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js:
```diff
@@ -256,6 +277,13 @@ export default compose(
+		// TODO (blixt): A unit-testable function would be nice, once there are
+		// multiple blockType.name values that have this feature.
+		// TODO (blixt): Check out /block-switcher/index.js for web's implementation
+		const isTransformEnabled = blockName === 'core/paragraph';
```
I think that would be nice! I don't see a way to access that definition, though - the block definition isn't exported from that package. How can I reference that definition?

blixt-io commented on packages/block-editor/src/components/block-mobile-toolbar/block-actions-menu.native.js:
```diff
@@ -256,6 +277,13 @@ export default compose(
+		// NOTE (blixt): To start, we're enabling the Transform menu just for
+		// Paragraph blocks - we can expand this list later-on!
+		// TODO (blixt): A unit-testable function would be nice, once there are
+		// multiple blockType.name values that have this feature.
```
I've now added the function and some unit tests!

blixt-io commented:
@chipsnyder I've made some updates since your last review:
- pulled in changes from the other two PRs, now it's just a single PR.
- extracted the "what transforms are available for this block in the mobile editor?" function: `getMobileBlockTransformations`, and added unit tests for the function
- restricted the implementation to only enable transforms on `core/paragraph` blocks; we can lift this restriction in the future! I figured that since there are probably design-requirements for other block-to-block transformations, we might want to take it slow on this ticket.

Mind taking another look when you're back in? Thank you!

blixt-io commented:
(Huh... I think GitHub added a bunch of reviewers automatically to this review! Sorry for the spam, everybody - my intent is to review things with @chipsnyder. I've marked this PR as a "draft" to hopefully remove it from your GitHub queues!)

gziolo commented on packages/blocks/src/api/factory.js:
```diff
@@ -344,6 +344,43 @@ export function getPossibleBlockTransformations( blocks ) {
+ * @return {Array} Block types that the block argument can be transformed to in
+ * the mobile app.
+ */
+export function getMobileBlockTransformations( block ) {
```
Have you considered hooking into block registration on mobile and picking only subset of transforms instead? You could either write a filter that updates the list of registered transforms with:

https://github.com/WordPress/gutenberg/blob/master/docs/designers-developers/developers/filters/block-filters.md#blocksregisterblocktype

or you could extract existing transforms to their own files and override them with a native version of the file.

In general, having a mobile specific name of the same functionality is confusing here.

chipsnyder replied:
What @gziolo mentioned here is more of what I had in mind from our side conversation @blixt-io.

chipsnyder commented on packages/blocks/src/api/factory.js:
```diff
@@ -344,6 +344,43 @@ export function getPossibleBlockTransformations( blocks ) {
+		return [];
+	}
+
+	// NOTE (blixt): From what I can gather, we want to support most-any
+	// transform, except that we don't want to support transforming into a
+	// column or group. If the rules are more nuanced, we could have a mapping
+	// of disallowed transformations - but for now, let's do a simple blocklist!
+	const ineligibleDestinationBlocks = [ 'core/column', 'core/group' ];
```
> Want to ban specific block types from transforming to core/column.

This would align more with what we discussed on our side conversation that we might want to ban specific options.

chipsnyder replied:
We also probably only want to support our list of supported blocks which can be found here: https://github.com/WordPress/gutenberg/blob/47feef0a2aaaa38b34fb3525a35cfddaf032c719/packages/block-library/src/index.native.js#L195-L227

chipsnyder commented on packages/blocks/src/api/factory.js:
```diff
@@ -344,6 +344,43 @@ export function getPossibleBlockTransformations( blocks ) {
+export function getMobileBlockTransformations( block ) {
+	// NOTE (blixt): For now, we only support transforming *from* the paragraph
+	// block. We'll add others soon!
+	const eligibleOriginBlocks = [ 'core/paragraph' ];
```
I wonder if we can tie into the block registration to mark certain blocks as eligible for transformation.

chipsnyder commented:
Closing this PR as it's no longer being worked on
